pkg/controller/backupbucket/actuator.go

```diff
@@ -6,43 +6,164 @@ package backupbucket
 
 import (
 	"context"
+	"errors"
 
+	"cloud.google.com/go/storage"
 	"github.com/gardener/gardener/extensions/pkg/controller/backupbucket"
 	"github.com/gardener/gardener/extensions/pkg/util"
 	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
 	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/admission"
+	apisgcp "github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp"
 	"github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp/helper"
 	gcpclient "github.com/gardener/gardener-extension-provider-gcp/pkg/gcp/client"
 )
 
 type actuator struct {
 	backupbucket.Actuator
-	client client.Client
+	client           client.Client
+	gcpClientFactory gcpclient.Factory
 }
 
-func newActuator(mgr manager.Manager) backupbucket.Actuator {
+// NewActuator creates a new Actuator that manages BackupBucket resources.
+func NewActuator(mgr manager.Manager, gcpClientFactory gcpclient.Factory) backupbucket.Actuator {
 	return &actuator{
-		client: mgr.GetClient(),
+		client:           mgr.GetClient(),
+		gcpClientFactory: gcpClientFactory,
 	}
 }
 
-func (a *actuator) Reconcile(ctx context.Context, _ logr.Logger, bb *extensionsv1alpha1.BackupBucket) error {
-	storageClient, err := gcpclient.NewStorageClientFromSecretRef(ctx, a.client, bb.Spec.SecretRef)
+func (a *actuator) Reconcile(ctx context.Context, logger logr.Logger, bb *extensionsv1alpha1.BackupBucket) error {
+	logger = logger.WithValues("name", bb.Name)
+	logger.Info("Starting reconciliation for BackupBucket")
+
+	storageClient, err := a.gcpClientFactory.Storage(ctx, a.client, bb.Spec.SecretRef)
 	if err != nil {
+		logger.Error(err, "Failed to create storage client")
 		return util.DetermineError(err, helper.KnownCodes)
 	}
 
-	return util.DetermineError(storageClient.CreateBucketIfNotExists(ctx, bb.Name, bb.Spec.Region), helper.KnownCodes)
+	var backupBucketConfig *apisgcp.BackupBucketConfig
+	if bb.Spec.ProviderConfig != nil {
+		backupBucketConfig, err = admission.DecodeBackupBucketConfig(serializer.NewCodecFactory(a.client.Scheme(), serializer.EnableStrict).UniversalDecoder(), bb.Spec.ProviderConfig)
+		if err != nil {
+			logger.Error(err, "Failed to decode provider config")
+			return err
+		}
+	}
+
+	attrs, err := storageClient.Attrs(ctx, bb.Name)
+	if err != nil && !errors.Is(err, storage.ErrBucketNotExist) {
+		logger.Error(err, "Failed to fetch bucket attributes")
+		return util.DetermineError(err, helper.KnownCodes)
+	}
+
+	if errors.Is(err, storage.ErrBucketNotExist) {
+		attrs, err = createBucket(ctx, storageClient, bb, backupBucketConfig, logger)
+		if err != nil {
+			return err
+		}
+
+	} else if backupBucketConfig != nil && isUpdateRequired(attrs, backupBucketConfig) {
```

Comment:
I've moved nil check inside func `isUpdateRequired`  as part of c5c15f9afe942ba8a3e1c0fd658312f0b469b5be

Reply:
>Additional thoughts that I'm writing down here:
>
>During development, it would be good if it is possible to remove immutability through this configuration. As it stands now, the validating webhook would disallow values lesser than 24h. Would it be possible to make the webhook ignore such resources through an explicit annotation that is set on the resource?

Since other providers are also enforcing a 1-day minimum period, let's proceed with 24 hours as the standard.

Reply:
Sure. Thanks!